### PR TITLE
Use stateless getLogs() rpc instead of event filters across codebase

### LINKED ISSUE #2291: Event filters can sometimes disappear and cause 'filter not found' error
User experiences the following after some period of time:
```
2020-09-23T14:29:04+0000 [twisted.internet.defer#critical] Unhandled error in Deferred:
2020-09-23T14:29:04+0000 [twisted.internet.defer#critical] 
    Traceback (most recent call last):
      File "/usr/local/lib/python3.6/dist-packages/twisted/internet/base.py", line 1283, in run
        self.mainLoop()
      File "/usr/local/lib/python3.6/dist-packages/twisted/internet/base.py", line 1292, in mainLoop
        self.runUntilCurrent()
      File "/usr/local/lib/python3.6/dist-packages/twisted/internet/base.py", line 913, in runUntilCurrent
        call.func(*call.args, call.kw)
      File "/usr/local/lib/python3.6/dist-packages/twisted/internet/task.py", line 239, in call
        d = defer.maybeDeferred(self.f, *self.a, self.kw)
    --- <exception caught here> ---
      File "/usr/local/lib/python3.6/dist-packages/twisted/internet/defer.py", line 151, in maybeDeferred
        result = f(*args, **kw)
      File "/usr/local/lib/python3.6/dist-packages/nucypher/utilities/prometheus/metrics.py", line 147, in collect_prometheus_metrics
        collector.collect()
      File "/usr/local/lib/python3.6/dist-packages/nucypher/utilities/prometheus/collector.py", line 77, in collect
        self._collect_internal()
      File "/usr/local/lib/python3.6/dist-packages/nucypher/utilities/prometheus/collector.py", line 303, in _collect_internal
        events = self.event_filter.get_new_entries()
      File "/usr/local/lib/python3.6/dist-packages/web3/_utils/filters.py", line 159, in get_new_entries
        log_entries = self._filter_valid_entries(self.web3.eth.getFilterChanges(self.filter_id))
      File "/usr/local/lib/python3.6/dist-packages/web3/eth.py", line 493, in getFilterChanges
        RPC.eth_getFilterChanges, [filter_id],
      File "/usr/local/lib/python3.6/dist-packages/web3/manager.py", line 153, in request_blocking
        raise ValueError(response["error"])
    builtins.ValueError: {'code': -32000, 'message': 'filter not found'}
```

Some interesting side notes:
- https://github.com/ethereum/web3.py/issues/923
- https://github.com/ethereum/web3.py/pull/732

### PR DESCRIPTION
**Pull request by derekpierre**
This facilitates the use of http(s) for infura to get events as opposed to getting the following error message:
```
ValueError: {'code': -32601, 'message': 'The method eth_newFilter does not exist/is not available’}
```

Fixes #2291 .
Related to #1670

### CODE REVIEW

File: nucypher/blockchain/eth/events.py [hunk 1]
```diff
@@ -100,13 +97,15 @@ def __init__(self,
             raise ValueError(f"Contract agent must be provided")
         if not event_name:
             raise ValueError(f"Event name must be provided")
+        if not from_block:
```
Comment (fjarri) [inline nucypher/blockchain/eth/events.py:100]: It feels strange that you have to check that for an argument that's positional, and explicitly typed as `int`. Also in the only call to `ContractEventsThrottler.__init__()` this value comes from `estimate_block_number_for_period` which does not return `None` either (on a side note, it returns `BlockNumber`, so perhaps that's what the type should be here).

Same applies to the other two checks, really.

File: nucypher/blockchain/eth/events.py [hunk 2]
```diff
@@ -100,13 +97,15 @@ def __init__(self,
...
+            raise ValueError(f"From block must be provided")
 
         self.event_filter = agent.events[event_name]
         self.from_block = from_block
         self.to_block = to_block if to_block else agent.blockchain.client.block_number
```
Comment (fjarri) [inline nucypher/blockchain/eth/events.py:105]: Can `to_block` be 0?

File: nucypher/blockchain/eth/events.py [hunk 3]
```diff
@@ -100,13 +97,15 @@ def __init__(self,
...
         # validity check of block range
-        if to_block <= from_block:
-            raise ValueError(f"Invalid block range provided ({from_block} - {to_block})")
+        if self.to_block <= self.from_block:
+            raise ValueError(f"Invalid block range provided ({self.from_block} - {self.to_block})")
```
Comment (fjarri) [inline nucypher/blockchain/eth/events.py:108]: May be better reported as "Final block must be greater than the starting block" or something.

Comment (fjarri) [reply nucypher/blockchain/eth/events.py:108]: Also, while it's not clear from `web3` documentations, [this SE link](https://ethereum.stackexchange.com/questions/78620/is-the-range-of-a-filterquery-inclusive-up-to-toblock) seems to indicate that the range is inclusive. So `to_block == from_block` is still a valid range.

File: nucypher/utilities/prometheus/collector.py [hunk 4]
```diff
@@ -309,9 +311,22 @@ def initialize(self, metrics_prefix: str, registry: CollectorRegistry) -> None:
...
+        events_throttler = ContractEventsThrottler(agent=self.contract_agent,
+                                                   event_name=self.event_name,
+                                                   from_block=from_block,
+                                                   to_block=to_block,
+                                                   **self.filter_arguments)
+        for event_record in events_throttler:
+            self._event_occurred(event_record.raw_event)
+
+        # update last block for the next round
+        self.filter_last_block_checked = to_block
```
Comment (derekpierre) [inline nucypher/utilities/prometheus/collector.py:329]: If the `to_block` for the query range is indeed inclusive (see @fjarri 's comment - https://github.com/nucypher/nucypher/pull/2322/files#r497823819)- this makes me think that this line would need to change to be:

```
self.filter_last_block_checked = to_block + 1
```

to ensure that the events for the same block isn't repeated during the next iteration. If so, other changes will need to be made to accommodate for this. Interested to get thoughts here?

File: nucypher/blockchain/eth/events.py (hunk 2 above)
Comment (fjarri) [reply nucypher/blockchain/eth/events.py:98]: Then perhaps it's better to check if `to_block is None` explicitly, because 0 would evaluate to `False` too.

Comment (derekpierre) [reply nucypher/blockchain/eth/events.py]: ✔️

File: nucypher/utilities/prometheus/collector.py (hunk 4 above)
Comment (derekpierre) [reply nucypher/utilities/prometheus/collector.py]: Thanks for pointing this out @fjarri! 😅 

The range is indeed inclusive, which may lead to duplicate events being obtained. It has now been fixed.

Here is an easy CLI test I ran locally to confirm:
```
derek @ Dereks-MacBook-Pro-2 ~/Documents/Github/repos/forks/derek/nucypher
 [20] → nucypher status events --network ibex --provider https://rinkeby.infura.io/v3/1e738a40d1b24d01b6bc7fb10fd9af3c --contract-name StakingEscrow --event-name CommitmentMade
Reading Latest Chaindata...
Showing events from block 7290043 to latest

--------- StakingEscrow Events ---------

CommitmentMade:
...
15000000000000000000000, block_number: 7292850

```

```
(nucypher) (base)
derek @ Dereks-MacBook-Pro-2 ~/Documents/Github/repos/forks/derek/nucypher
 [21] → nucypher status events --network ibex --provider https://rinkeby.infura.io/v3/1e738a40d1b24d01b6bc7fb10fd9af3c --contract-name StakingEscrow --event-name CommitmentMade --from-block 7292849 --to-block 7292850
Reading Latest Chaindata...
Showing events from block 7292849 to 7292850

--------- StakingEscrow Events ---------

CommitmentMade:
  - (EventRecord) staker: 0x488998c6dc3bA8d0Aee0A1952290d751C85eD730, period: 18537, value: 15000000000000000000000, block_number: 7292850
```

```
(nucypher) (base)
derek @ Dereks-MacBook-Pro-2 ~/Documents/Github/repos/forks/derek/nucypher
 [22] → nucypher status events --network ibex --provider https://rinkeby.infura.io/v3/1e738a40d1b24d01b6bc7fb10fd9af3c --contract-name StakingEscrow --event-name CommitmentMade --from-block 7292850 --to-block 7292851
Reading Latest Chaindata...
Showing events from block 7292850 to 7292851

--------- StakingEscrow Events ---------

CommitmentMade:
  - (EventRecord) staker: 0x488998c6dc3bA8d0Aee0A1952290d751C85eD730, period: 18537, value: 15000000000000000000000, block_number: 7292850
```

```
(nucypher) (base)
derek @ Dereks-MacBook-Pro-2 ~/Documents/Github/repos/forks/derek/nucypher
 [26] → nucypher status events --network ibex --provider https://rinkeby.infura.io/v3/1e738a40d1b24d01b6bc7fb10fd9af3c --contract-name StakingEscrow --event-name CommitmentMade --from-block 7292850 --to-block 7292850
Reading Latest Chaindata...
Showing events from block 7292850 to 7292850

--------- StakingEscrow Events ---------

CommitmentMade:
  - (EventRecord) staker: 0x488998c6dc3bA8d0Aee0A1952290d751C85eD730, period: 18537, value: 15000000000000000000000, block_number: 7292850
```

Comment (derekpierre) [reply nucypher/utilities/prometheus/collector.py]: ✔️

**Comment by cygnusv** [comment]
Needs a rebase to get rid of the CI error.

**Review by KPrasch — COMMENTED** [review]
Looks good - does this change the behavior of websocket providers at all?

**Comment by derekpierre** [comment]
@KPrasch it doesn't affect the behaviour of providers at all. It simply changes the web3 call we make for checking events i.e. that `eth_newFilter` is no longer used, and instead we use `eth_getLogs`.

**Comment by KPrasch** [comment]
Got that much - I am still trying to understand the impact of this change on the API since those two endpoints are used differently.

**Comment by derekpierre** [comment]
The filter is stateful so it stores state across calls, that's why you can do things like `get_new_entries()`, however, `eth_getLogs` is not. Example of filters going awry - https://github.com/ethereum/web3.py/issues/923.

**Review by KPrasch — APPROVED** [review]

**Review by tuxxy — APPROVED** [review]

**Comment by jdbertron** [comment]
Thanks for this.

**Pull request merged**